### PR TITLE
make Reactor toInputStream 50 times faster

### PR DESCRIPTION
**Pull request by mbaechler**
From 500 MiB/s to 10 GiB/s on my workstation

	It's due to the implementation of read(byte[], int off, int len)
	and some minor optimizations

### CODE REVIEW

**Comment by Arsnael** [comment]
```
[f48dc31aa52470dbb50b79310bf02ada49daa863] [INFO] There are 2 errors reported by Checkstyle 8.23 with checkstyle.xml ruleset.
[f48dc31aa52470dbb50b79310bf02ada49daa863] [ERROR] src/main/java/org/apache/james/util/ReactorUtils.java:[26,8] (imports) UnusedImports: Unused import - java.util.Spliterator.
[f48dc31aa52470dbb50b79310bf02ada49daa863] [ERROR] src/main/java/org/apache/james/util/ReactorUtils.java:[27,8] (imports) UnusedImports: Unused import - java.util.stream.Stream.
```
Let me help you with that

**Comment by chibenwa** [comment]
Can we create a JIRA for this too? Or reuse an existing one?

**Comment by Arsnael** [comment]
> Let me help you with that

Well permission denied... Please fix it then :)

**Comment by mbaechler** [comment]
Well, it's not exactly "refactoring". Refactoring is "changing the shape of the code without changing its behavior".
Here the goal is to change the behavior.

**Comment by remk** [comment]
```
[bc1b6c3aaab5b68925854b5c0a3cd25a2d9c9f44] [INFO] Results:
[bc1b6c3aaab5b68925854b5c0a3cd25a2d9c9f44] [INFO] 
[bc1b6c3aaab5b68925854b5c0a3cd25a2d9c9f44] [ERROR] Errors: 
[bc1b6c3aaab5b68925854b5c0a3cd25a2d9c9f44] [ERROR]   WebAdminUtilsTest.serverShouldBeAbletoStartConcurrently:38 » NotTerminated
```

test this please

**Comment by mbaechler** [comment]
```
[bc1b6c3aaab5b68925854b5c0a3cd25a2d9c9f44] [ERROR] Errors: 
[bc1b6c3aaab5b68925854b5c0a3cd25a2d9c9f44] [ERROR]   WebAdminUtilsTest.serverShouldBeAbletoStartConcurrently:38 » NotTerminated
```
It's unfortunate that this change, supposed to enhance the stability of the testsuite, is actually making the testsuite to fail

test this please

**Comment by chibenwa** [comment]
test this please

**Comment by Arsnael** [comment]
```
[bc1b6c3aaab5b68925854b5c0a3cd25a2d9c9f44] Exception in thread "Thread-1734" java.lang.RuntimeException: java.io.IOException: Failed to bind to /0.0.0.0:45597
[bc1b6c3aaab5b68925854b5c0a3cd25a2d9c9f44] 	at org.apache.james.webadmin.WebAdminServer.lambda$start$1(WebAdminServer.java:95)
[bc1b6c3aaab5b68925854b5c0a3cd25a2d9c9f44] 	at spark.Service.lambda$init$2(Service.java:582)
[bc1b6c3aaab5b68925854b5c0a3cd25a2d9c9f44] 	at java.base/java.lang.Thread.run(Thread.java:834)
```

**Comment by mbaechler** [comment]
test this please

**Comment by Arsnael** [comment]
test this please

File: server/container/util/src/test/java/org/apache/james/util/ReactorUtilsTest.java [hunk 1]
```diff
@@ -81,6 +84,37 @@ public Integer getCounter() {
 
     @Nested
     class ToInputStream {
+
+        @Test
+        void givenAFluxOf3BytesShouldConsumeT() throws IOException, InterruptedException {
```
Comment (rouazana) [inline server/container/util/src/test/java/org/apache/james/util/ReactorUtilsTest.java]: T?

File: server/container/util/src/test/java/org/apache/james/util/ReactorUtilsTest.java [hunk 2]
```diff
@@ -81,6 +84,37 @@ public Integer getCounter() {
...
+                .map(Bytes::toArray)
+                .map(ByteBuffer::wrap);
+
+            InputStream inputStream = ReactorUtils.toInputStream(source);
+
+            assertThat(inputStream).hasSameContentAs(new ByteArrayInputStream(bytes));
+        }
+
+        @Test
+        void givenALongFluxBytesShouldConsumeT() throws IOException, InterruptedException {
```
Comment (rouazana) [inline server/container/util/src/test/java/org/apache/james/util/ReactorUtilsTest.java]: T?

File: server/container/util/src/test/java/org/apache/james/util/ReactorUtilsTest.java [hunk 3]
```diff
@@ -150,7 +190,8 @@ void givenAnEmptyFluxShouldConsumeOnlyThePrefetch() throws IOException, Interrup
             inputStream.read(readBytes, 0, readBytes.length);
 
             assertThat(readBytes).contains(0, 0, 0, 0, 0);
-            Thread.sleep(200);
+            //make sure reactor is done with prefetch
+             Thread.sleep(200);
```
Comment (rouazana) [inline server/container/util/src/test/java/org/apache/james/util/ReactorUtilsTest.java]: indent

**Comment by Arsnael** [comment]
Merged

**Pull request closed without merging**